### ISSUE TITLE
Suggestion for a new feature - multiple rule sets

### ISSUE THREAD

**Issue by CapMan47** [dan-lovelace/word-replacer-max#52]
Hey, great extension, thanks for creating it)
I'd like to suggest a new feature - the ability to create multiple rule sets and switch between them. 
It would be handy to make different sites their own rule sets to replace, at least for me. 
I hope you'll consider my suggestion, thanks in advance)

**Comment by dan-lovelace** [comment]
Hey @CapMan47 -

Thanks for the suggestion, this would indeed be a nice improvement. Please allow a couple days for me to think about the best way to introduce something like this without causing too much interruption to the current way of things. I'll follow up in this thread once I have a solution in mind to make sure what I build out will meet your needs.

**Comment by dan-lovelace** [comment]
Alright @CapMan47, I'm starting to get an idea of how this could work. I could use a hand clarifying your original request before I provide further detail. You mention these two things:

> the ability to create multiple rule sets and switch between them

> make different sites their own rule sets to replace

It sounds in the first quote like you intend to group rules into sets and switch between them manually. The second quote then makes me think you're also looking to have rulesets applied automatically when you visit certain sites. So essentially two different features:

1. Ability to create groups of rules into rulesets - The user should be able to switch between rulesets at-will
2. Ability to manage domains in a way that allows automatically applying one or more rulesets depending on the current site - The user does not need to manually switch rulesets

Do these two items sound like what you're looking for?

**Comment by CapMan47** [comment]
It's nice that you don't forget even in the new year, I wish you wonderful New Year holidays) 
As for the request, I meant the first option - the ability to create different lists of rules and switch between them manually. I must say that the second one is also very interesting.

**Comment by dan-lovelace** [comment]
Happy New Year and holidays to you as well!

Thanks for the clarification, this should be fairly straightforward to introduce. Below are some details around how I see it working from a UI perspective. I'd be happy to hear any feedback, otherwise I'll get started on these updates right away.

---

## New UI components

- New setting on the **Options** tab named "Rulesets" (can be toggled ON/OFF)
- When enabled, a new tab **Rulesets** becomes available and the **Rules** tab changes slightly to allow adding/removing rulesets manually (one or more rulesets may be applied at any time). Also, each rule row gets a new option "Add to ruleset" to easily add to sets on-the-fly. I'm thinking a "Show all" toggle should exist that allows showing all rules or filtering the displayed rules by only the chosen set(s).
- On the **Rulesets** tab, the list of sets are displayed with the ability to add, edit, and delete them. Each one has a required "Name" field and users will add rules to the set using a similar modal component as used in the "Rules Export" feature. Regarding the second option above around domains, this might also be a good place to manage a list of domains where the set will automatically be applied. I'll consider adding this as part of the initial changes but may choose to defer it to a later time if I come across any complications. One tricky thing here is how to properly display to the user when rules are being overridden based on ruleset domain settings 🤔
- A built-in ruleset named "Default" will always exist and cannot be edited. This will serve as a "fallback" mechanism when no rulesets are currently applied.
- On the **Rules** tab with one or more rulesets applied, a new "Add to set" dropdown menu button is shown beside the "Add rule" button where the user can choose a set in which to create the rule. The current "Add rule" button will remain accessible and will create a new rule in the global scope (no set).

## Other considerations

- Ability to import & export rulesets in the sharing features
- All rules' "active" state will be managed independently at both the global and ruleset scope. This allows users to enable/disable rules within sets while having their globally-scoped enabled/disabled state kept intact. It aims to solve for situations where a user may want to temporarily disable a rule in a set but keep it enabled globally. Instead of having to re-add the rule via the modal, they would simply re-enable it at a later time.

**Comment by CapMan47** [comment]
Thank you)
Just wanted to clarify, will the user be able to switch ruleset within the same site? Will it be possible to enable/disable rulesetst independently? 
To give you an example of my use case: I work with text in English and Spanish, but on the same site (online notepad) and for both languages I would like to switch between rulesets as needed.

**Comment by dan-lovelace** [comment]
That's correct, you will be able to switch between them any time. I've created a small demo video highlighting what I have in place so far with the ruleset switcher. None of the rules are actually filtered yet, this demo is meant to show off the switching component only to give you an idea of how it could work. Hopefully it aligns with your expectation but let me know if I'm missing something.

https://github.com/user-attachments/assets/d8ab7c27-d05b-4dd5-a62c-8c57ca3e21f9

**Comment by CapMan47** [comment]
Yeah, that's exactly how I envisioned it. Thanks for the work and feedback, I'll wait for the release)

**Comment by dan-lovelace** [comment]
Hey @CapMan47 -

This feature is just about ready to ship. I have a few loose ends to tie up tomorrow and will publish to the stores after. The version you're looking for is v0.11.0 which should be available in the stores in the next few days.

After you've received it, don't hesitate to reach back out, either in this thread or elsewhere, if you have any questions or need anything updated to work differently. Thanks for your patience and feedback along the way, cheers!

**Comment by dan-lovelace** [comment]
Heads up @CapMan47, I found a bug with the groups functionality that has the potential of permanently overwriting saved rules. An updated version is on its way to the stores but I wanted to warn you personally in hopes of avoiding it. Please leave the groups feature disabled on the Options page until v0.13.0 is available for your browser. Apologies for the inconvenience.

**Comment by dan-lovelace** [comment]
Hey @CapMan47 - Just wanted to let you know the rulesets/groups feature has been reintroduced to the extension as of version 0.15.1 and the previous bug fixed. Thanks for bearing with me through the tough rollout of this feature!
